add icon support for Windows

maphew opened the pull request:
Added `icon='path/to/file.ico')` keyword argument to `create_shortcut()` and related. 

It works on Windows for me ([ref][0]) but I won't be surprised if it's not a complete solution. If it proves to be good enough, Linux/Mac support should be straightforward following same pattern I used.

[0]: https://github.com/leo-editor/leo-editor/blob/desktop-links/leo/scripts/add-desktop-links.leox

mhw-at-yg committed:
does simple hard coded icon path work? Yes

maphew committed:
WIP: tracking down complete Icon parameter exec path

We're losing the value somewhere before `_create_shortcut_win()` is
called.

```
$ leo-m --script leo\scripts\add-desktop-links.leox

Leo 6.1-devel, devel branch, build 3cd45df95c
2019-08-08 12:57:29 -0700
Running add-desktop-links...
create_desktop_shortcut() D:\code-maphew\leo-editor\leo\Icons\LeoApp.ico
create_shortcut() D:\code-maphew\leo-editor\leo\Icons\LeoApp.ico
_create_shortcut_win(): None
//_create_shortcut_win %SystemRoot%\explorer.exe,0
//create_shortcut() D:\code-maphew\leo-editor\leo\Icons\LeoApp.ico
create_shortcut() None
_create_shortcut_win(): None
//_create_shortcut_win %SystemRoot%\System32\imageres.dll,11
//create_shortcut() None
_create_shortcut_win(): None
//_create_shortcut_win %SystemRoot%\System32\imageres.dll,11
_create_shortcut_win(): None
//_create_shortcut_win %SystemRoot%\System32\imageres.dll,11
_create_shortcut_win(): None
//_create_shortcut_win %SystemRoot%\System32\imageres.dll,11
_create_shortcut_win(): None
//_create_shortcut_win %SystemRoot%\System32\imageres.dll,11
Created shortcut: Leo Editor - C:\Users\matt\Desktop\Leo Editor.lnk
Created shortcut: Leo Editor - C:\Users\matt\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Leo Editor.lnk
```

maphew committed:
more debug code scattering. This isn't right way to solve this...

maphew committed:
fixed losing icon path! (line 87)

Bug was that we turned 'icon' from optional arg to fixed position, causing
next function to use it for 'folder' arg instead of 'icon'. Fix is make
sure is optional: `icon=icon`.

(Todo: as near as I can tell `_create_shortcut_win()` doesn't use 'Folder'
arg at all. Add use of it or remove.)

maphew committed:
Merge branch 'icons-win' of https://github.com/maphew/shortcutter into icons-win

maphew committed:
bug fix: icon is keyword arg, not fixed positional

Also:

 + added icon notes to doc strings
 - removed debug prints

maphew committed:
1 step towards remembering Terminal files created

Creating global dict of paths to files created because this func
returning true/false is relied on elsewhere, and upstream is reluctant
so far to change that. (Also opens possibility of having other functions
add ot the dict, and saving the paths persistently.)

maphew committed:
WIP: linux support for icons (untested)

The calls between Windows and Linux functions are slightly different,
and were before I started messing about. Be good to ensure the args are
the same across all of the functions regardless of platform before a
general release.

kiwi0fruit commented:
Looks almost OK. But I guess it needs stubs for mac and linux extra arguments. I also think that there is no need to return icon path in a tuple. And what is this:

```
global linkd
```
Why? Is it necessary?

kiwi0fruit commented:
There is also a typo on Linux:

```
        return self._create_shortcut_linux(shortcut_name, target_path, shortcut_directory,
                                           '[Desktop Entry]\n' +
                                           'Name={}\n'.format(shortcut_name) +
                                           'Type=Application\n' +
                                           'Path={}\n'.format(target_path) +
                                           'Exec=xdg-open "{}"\n'.format(target_path) +
                                           'Icon=system-file-manager.png\n')
                                           'Icon={}\n',format(icon))
```

I guess I should add a travis at least the same as in conda forge build.

kiwi0fruit commented:
I added travis for windows, mac and linux. So it should check the next your commit here (I guess).

It checks creation of the shortcuts (see [.travis.yml](https://github.com/kiwi0fruit/shortcutter/blob/master/.travis.yml#L52)). By the way: do you have any ideas how to check there if the shourtcuts themselves work?

maphew commented:
The linux part is partial. I checked in on my way out the door since I won't be back at that computer for several days. It should have been in a branch to avoid getting in the middle of the pull request and causing confusion.

The idea of global `linkd` is so that the path to the terminal shortcuts could be captured and reused elsewhere without changing the function's current return values of True/False. (In #12 you didn't sound warm to that prospect.) I'm not committed to this approach.

Travis is a great idea. Not sure about how to test if a shortcut is viable or not. Maybe a small script that emits "Hello world, I'm here!"? Install shortcutter, run it against helloworld, then attempt to run `path/to/desktop/Hello World.desktop` ?

I found 1 Stack Q about running an Electron app in Travis and capturing screenshots that might be fruitful: https://stackoverflow.com/questions/42813924/test-electron-app-with-spectron-and-travis

Thanks for reviewing! I'll cleanup the half done linux work and move it out of the way.

kiwi0fruit commented:
linkd doen't look right for the purpose mentuoned. And I guess it should be separate PR anyway.

So if you remove  linkd, fix linux typo, add macos stubs (so there won't be unexpected arguments for them) and remove (or justify) passing icon path in return then I'll test and merge.

maphew committed:
Revert "1 step towards remembering Terminal files created"

This reverts commit 9e9a85c4e39cee52ac559789e673d3a0e544f798.

maphew committed:
icons: make arg calls match across all OS .py's

linux.py and macos.py not tested by me. Let's see what TravisCI says in
a few minutes.

maphew commented:
I reverted the half-baked linux icon support and pushed a change which makes the argument calls across all three OS .py match each other. There's no support within mac and linux functions to use the new icon parameter yet.

macos.py: has it's own `create_shortcut()`, I presume overwriting or taking precedence over same name function in base.py. base.py has keyword `shortcut_name=None` instead of the positional arg in macos. I left this mis-match alone.

Travis is happy with all 3 builds (Windows is 6 times slower! Good grief MS.)

kiwi0fruit commented:
Closing for now as I guess nobody needs it now. Neither me nor maphew. Changing output signature - adding extra None to the return tuple is of no use here. But I won't fix it...

kiwi0fruit commented:
CLI interface is also absent...
